packages/gatsby/src/utils/source-nodes.ts

```diff
@@ -116,15 +173,18 @@ export default async ({
 
   await getDataStore().ready()
 
-  // We only warn for plugins w/o nodes and delete stale nodes on the first sourcing.
-  if (isInitialSourcing) {
-    const state = store.getState()
-    const nodes = getDataStore().iterateNodes()
+  // We only warn for plugins w/o nodes and delete stale nodes on the first sourceNodes call of the current process.
+  if (isInitialSourceNodesOfCurrentNodeProcess) {
+    isInitialSourceNodesOfCurrentNodeProcess = false
 
-    warnForPluginsWithoutNodes(state, nodes)
+    warnForPluginsWithoutNodes()
 
-    deleteStaleNodes(state, nodes)
-    isInitialSourcing = false
+    if (
+      // if this is the very first source and no types existed before this sourceNodes run, there's no need to check for stale nodes. They wont be stale because they were just created. Only check for stale nodes in node types that never existed before.
```

Comment:
this occurred to me while making the latest changes. There's no need to check stale nodes for types that didn't exist before the most recent `sourceNodes` run. Anything that didn't exist before is not stale because it was just created. This should be a speed boost for all sites/source plugins even if they are still calling `touchNodes`